jlm/rvsdg/region.cpp

```diff
@@ -360,47 +361,120 @@ region::NumRegions(const jlm::rvsdg::region & region) noexcept
-  auto identationChar = '-';
+  static const char identationChar = '-';
+  static const char annotationSeparator = ' ';
+  static const char labelValueSeparator = ':';
+
+  auto identationString = std::string(identationDepth, identationChar);
+  auto regionString =
+      region.IsRootRegion() ? "RootRegion" : util::strfmt("Region[", region.index(), "]");
+  auto regionAnnotationString =
+      GetAnnotationString(&region, annotationMap, annotationSeparator, labelValueSeparator);
 
   // Convert current region to a string
-  if (region.IsRootRegion())
-  {
-    subTree = "RootRegion\n";
-    identationDepth += 1;
-  }
-  else if (region.node()->nsubregions() != 1)
-  {
-    auto indentationString = std::string(identationDepth, identationChar);
-    subTree += util::strfmt(indentationString, "Region[", region.index(), "]\n");
-    identationDepth += 1;
-  }
+  stream << identationString << regionString << regionAnnotationString << '\n';
+  identationDepth += 1;
 
   // Convert the region's structural nodes with their subregions to a string
-  for (const auto & node : region.nodes)
+  for (auto & node : region.nodes)
   {
     if (auto structuralNode = dynamic_cast<const rvsdg::structural_node *>(&node))
     {
-      auto identationString = std::string(identationDepth, identationChar);
       auto nodeString = structuralNode->operation().debug_string();
-      subTree += util::strfmt(identationString, nodeString, '\n');
+      auto annotationString = GetAnnotationString(
+          structuralNode,
+          annotationMap,
+          annotationSeparator,
+          labelValueSeparator);
+      stream << identationString << nodeString << annotationString << '\n';
 
       for (size_t n = 0; n < structuralNode->nsubregions(); n++)
       {
-        subTree += ToTree(*structuralNode->subregion(n), identationDepth + 1);
+        ToTree(*structuralNode->subregion(n), annotationMap, identationDepth + 1, stream);
       }
     }
   }
+}
+
+std::string
+region::GetAnnotationString(
+    const void * key,
+    const util::AnnotationMap & annotationMap,
+    char annotationSeparator,
+    char labelValueSeparator)
+{
+  if (!annotationMap.HasAnnotations(key))
+    return "";
+
+  auto & annotations = annotationMap.GetAnnotations(key);
+  return ToString(annotations, annotationSeparator, labelValueSeparator);
+}
+
+std::string
+region::ToString(
+    const std::vector<util::Annotation> & annotations,
+    char annotationSeparator,
+    char labelValueSeparator)
+{
+  std::stringstream stream;
+  for (auto & annotation : annotations)
+  {
+    auto annotationString = ToString(annotation, labelValueSeparator);
+    stream << annotationSeparator << annotationString;
+  }
+
+  return stream.str();
+}
+
+std::string
+region::ToString(const util::Annotation & annotation, char labelValueSeparator)
+{
+  std::string value;
+  if (annotation.HasValueType<std::string_view>())
+  {
+    value = util::strfmt(annotation.Value<std::string_view>());
```

Comment:
Do we need strfmt to convert a `string_view` to a string? I would think it is implicit?

On another note, is it maybe a bit limiting to only have `string_view` as the string value, in case some strings need to be heap allocated. I have always wished for a string type that can both represent owned and non-owned strings.

Reply:
Removed the strfmt. I also changed the string value type from `std::string_view` to `std::string`